include/sys_clock.h

```diff
@@ -1,248 +1,109 @@
-
+#ifdef __cplusplus
+extern "C" {
 #endif
 
-static ALWAYS_INLINE s32_t z_ms_to_ticks(s32_t ms)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-
-#ifdef _NEED_PRECISE_TICK_MS_CONVERSION
-	int cyc = sys_clock_hw_cycles_per_sec();
-
-	/* use 64-bit math to keep precision */
-	return (s32_t)ceiling_fraction((s64_t)ms * cyc,
-		((s64_t)MSEC_PER_SEC * cyc) / CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+#ifdef CONFIG_SYS_TIMEOUT_64BIT
+typedef u64_t k_ticks_t;
 #else
-	/* simple division keeps precision */
-	s32_t ms_per_tick = MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-
-	return (s32_t)ceiling_fraction(ms, ms_per_tick);
+typedef u32_t k_ticks_t;
 #endif
 
+#ifdef CONFIG_SYS_TIMEOUT_LEGACY_API
+/* Fallback API where timeouts are still 32 bit millisecond counts */
+typedef k_ticks_t k_timeout_t;
+#define K_TIMEOUT_TICKS(t) (t)
+#define K_TIMEOUT_GET(t) (t)
+#define K_TIMEOUT_EQ(a, b) ((a) == (b))
 #else
-	__ASSERT(ms == 0, "ms not zero");
-	return 0;
+/* New API going forward: k_timeout_t is an opaque struct */
+typedef struct { k_ticks_t ticks; } k_timeout_t;
+#define K_TIMEOUT_TICKS(t) ((k_timeout_t){ (k_ticks_t)t })
+#define K_TIMEOUT_GET(t) ((t).ticks)
+#define K_TIMEOUT_EQ(a, b) ((a).ticks == (b).ticks)
 #endif
-}
 
-static inline u64_t __ticks_to_ms(s64_t ticks)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return (u64_t)ticks * MSEC_PER_SEC /
-	       (u64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-#else
-	__ASSERT(ticks == 0, "ticks not zero");
-	return 0ULL;
-#endif
-}
+struct _timeout;
+typedef void (*_timeout_func_t)(struct _timeout *t);
 
-/*
- * These are only currently used by k_usleep(), but they are
- * defined here for parity with their ms analogs above. Note:
- * we don't bother trying the 32-bit intermediate shortcuts
- * possible with ms, because of the magnitudes involved.
- */
+struct _timeout {
+	sys_dnode_t node;
+	k_ticks_t dticks;
+	_timeout_func_t fn;
+};
 
-static inline s32_t z_us_to_ticks(s32_t us)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return (s32_t) ceiling_fraction(
-		(s64_t)us * sys_clock_hw_cycles_per_sec(),
-		((s64_t)USEC_PER_SEC * sys_clock_hw_cycles_per_sec()) /
-		CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+// FIXME: these are the new APIs, need doxygen.
+#ifdef CONFIG_SYS_TIMEOUT_64BIT
+#define K_TIMEOUT_MS(t) K_TIMEOUT_TICKS(k_ms_to_ticks_ceil64(t))
+#define K_TIMEOUT_US(t) K_TIMEOUT_TICKS(k_us_to_ticks_ceil64(t))
+#define K_TIMEOUT_CYC(t) K_TIMEOUT_TICKS(k_cyc_to_ticks_ceil64(t))
 #else
-	__ASSERT(us == 0, "us not zero");
-	return 0;
+#define K_TIMEOUT_MS(t) K_TIMEOUT_TICKS(k_ms_to_ticks_ceil32(t))
+#define K_TIMEOUT_US(t) K_TIMEOUT_TICKS(k_us_to_ticks_ceil32(t))
+#define K_TIMEOUT_CYC(t) K_TIMEOUT_TICKS(k_cyc_to_ticks_ceil32(t))
 #endif
-}
 
-static inline s32_t __ticks_to_us(s32_t ticks)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return (s32_t) ((s64_t)ticks * USEC_PER_SEC /
-	       (s64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC);
-#else
-	__ASSERT(ticks == 0, "ticks not zero");
-	return 0;
+// Uptime API: specify a timeout to happen at (i.e. immediately after,
```

Comment:
i'm a bit worried about adding new feature (absolute timings) working only with 64bit. It is performance and rom footprint penalty. I tried simple function (adding return u64+u32 vs. return u32 + u32) and u64 version took twice more (20 cycles vs 40 cycles). 

I was then wondering if maybe it would be better to use term 'absolute' instead of uptime and define `k_timeout_t` as
```
typedef struct {
    k_ticks_t relative: 1;
    k_ticks_t ticks: K_TICKS_BITS - 1;
} k_timeout_t;
```
For devices with 32k system clock frequency 2^31 is 18 hours so that's more then enough (in most of the cases).